### PR TITLE
Typed Setters for raise_on_nil_write Props

### PR DESCRIPTION
**Pull request by ezra-stripe**
Make it so setters for raise_on_nil_write props no longer allow nil.

### Motivation

We currently enforce this via runtime checks, but it would be more helpful if the signature correctly implied the use case.

### Test plan

See included automated tests.

### CODE REVIEW

File: gems/sorbet-runtime/test/types/struct.rb [hunk 1]
```diff
@@ -44,4 +44,38 @@ class Opus::Types::Test::StructValidationTest < Critic::Unit::UnitTest
...
+      m.foo = nil
+    end
+  end
+
+  # Failing - missing required prop
+  it "raises a TypeError omitting raise_on_nil_write prop on initialization" do
+    klass = Class.new(T::Struct) do
+      prop :foo, T.nilable(String), raise_on_nil_write: true
+    end
+    klass::new
```
Comment (jez) [inline gems/sorbet-runtime/test/types/struct.rb:63]: Can you write `klass.new` instead?

**Comment by ezra-stripe** [comment]
I'm pretty happy with this change now, but please take a look and let me know if you want any specific changes.
It also still needs to go through internal testing against pay-server, but I'm not able to run those tests due to permissioning issues on the llvm repo.

**Comment by jez** [comment]
I'm on leave this week; assigning someone else

File: test/testdata/rewriter/prop_raise_on_nil_write.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+# typed: strict
+class Thing
+  include T::Props
+  prop :foo, T.nilable(String), raise_on_nil_write: true
+end
```
Comment (elliottt) [inline test/testdata/rewriter/prop_raise_on_nil_write.rb:5]: Could you add some cases that show the behavior of writes with nil values?

File: rewriter/Prop.cc [hunk 3]
```diff
@@ -337,7 +351,13 @@ vector<ast::ExpressionPtr> processProp(core::MutableContext ctx, PropInfo &ret,
 
     // Compute the setter
     if (!ret.isImmutable) {
-        auto setType = ASTUtil::dupType(ret.type);
+        ast::ExpressionPtr setType;
+        // When we have raise_on_nil_write, the setType should be the non-nilable underlying type
+        if (ret.raiseOnNilWrite && isTNilable(ret.type)) {
+            setType = ASTUtil::dupType(getUnderlyingType(ret.type));
```
Comment (elliottt) [inline rewriter/Prop.cc:357]: Could you add a test case that shows that this won't detect `T.any(NilClass, <other-type>)`? That's the interpretation of `T.nilable(<other-type>)`, but this implementation is sensitive to the use of `T.nilable`.

**Review by elliottt — CHANGES_REQUESTED** [review]
A few requests for additional test cases, but otherwise this makes sense to me!

File: test/testdata/rewriter/prop_raise_on_nil_write.rb (hunk 2 above)
Comment (ezra-stripe) [reply test/testdata/rewriter/prop_raise_on_nil_write.rb]: Added a few tests, both in the rewriter and also in the runtime for struct.

File: gems/sorbet-runtime/test/types/struct.rb [hunk 4]
```diff
@@ -44,4 +44,65 @@ class Opus::Types::Test::StructValidationTest < Critic::Unit::UnitTest
...
+
+  it "allows setting non-nil values with T.any(NilClass, U) with raise_on_nil_write props" do
+    klass = Class.new(T::Struct) do
+      prop :foo, T.any(NilClass, String), raise_on_nil_write: true
+    end
+    m = klass.new
+    m.foo = 'foo'
+  end
+
+  it "allows setting nil values with T.nilable(NilClass) with raise_on_nil_write props, which is fine maybe?" do
```
Comment (ezra-stripe) [inline gems/sorbet-runtime/test/types/struct.rb:101]: This is a bit odd, but I'm not sure what the desired behavior would be in any case, so I suppose it is fine as is.

**Comment by ezra-stripe** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_J5PtpMh6D7L5vd
→ https://go/builds/bui_J5Pt9z30461rSz

**Comment by ezra-stripe** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_J6523xAqRbtRD9
→ https://go/builds/bui_J652faVsV38ela

**Comment by ezra-stripe** [comment]
The new approach appears to generally work, and I'm now fixing pieces in pay-server so we can get a green build. More things might come up, we'll see.

**Comment by ezra-stripe** [comment]
Now that we got it mostly working, we've decided to pursue removing `raise_on_nil_write` rather than make these changes. ;)

**Pull request closed without merging**